File: constgen.zig
const std = @import("std");

fn vecOps(size: usize, comptime Val: type) type {
    return struct {
        // fn dot(a: []const Val, b: []const Val) Val {
        fn dot(a: [size]Val, b: [size]Val) Val {
            var result: Val = 0;
            for (a) |val_a, index| {
                result += val_a * b[index];
            }
            return result;
        }

        fn norm(a: [size]Val) Val {
            return @sqrt(dot(a, a));
        }
    };
}

const vec2f32 = vecOps(2, f32);
// export const norm2f32 = vec2f32.norm;

pub fn main() void {
    const a = [_]f32{ 1.5, 2 };
    std.debug.print("norm: {}\n", .{vec2f32.norm(a)});
}

File: constgen.rs
use num_traits::Float;
use std::{array::IntoIter};  //, mem::size_of};

fn dot<Val: Float, const SIZE: usize>(
    a: [Val; SIZE],
    b: [Val; SIZE],
) -> Val {
    // a.iter()
    //     .zip(b)
    //     .map(|(x, y)| *x * *y)
    //     .fold(Val::zero(), |sum, xy| sum + xy)
    IntoIter::new(a)
        .zip(IntoIter::new(b))
        .map(|(x, y)| x * y)
        .fold(Val::zero(), |sum, xy| sum + xy)
}

fn norm<Val: Float, const SIZE: usize>(a: [Val; SIZE]) -> Val {
    dot(a, a).sqrt()
}

// pub fn norm2(x: f32, y: f32) -> f32 {
//     return norm([x, y]);
// }

fn main() {
    let a = [1.5, 2.0f32];
    println!("norm: {}", norm::<f32, 2>(a));
    // println!("sizes: {} {}", size_of::<[f32; 2]>(), size_of::<&[f32]>());
}

// -C opt-level=3

// use std::ops::*;

// trait Float: Copy + Sized + Add<Self, Output = Self> + Mul<Self, Output = Self> {
//     fn sqrt(self) -> Self;

//     fn zero() -> Self;
// }

// impl Float for f32 {
//     fn sqrt(self) -> f32 {
//         self.sqrt()
//     }

//     fn zero() -> f32 {
//         0f32
//     }
// }
